core/src/databases/database.rs

```diff
@@ -265,28 +291,25 @@ impl Table {
             }
         }
 
-        // Validate the tables schema, merge it if necessary and store it in the schema cache.
-        let table_schema = match truncate {
-            // If the new rows replace existing ones, we need to clear the schema cache.
-            true => TableSchema::from_rows(&rows)?,
-            false => match self.schema() {
-                // If there is no existing schema cache, simply use the new schema.
-                None => TableSchema::from_rows(&rows)?,
-                Some(existing_table_schema) => {
-                    // If there is an existing schema cache, merge it with the new schema.
-                    existing_table_schema.merge(&TableSchema::from_rows(&rows)?)?
-                }
-            },
-        };
-
-        store
-            .update_table_schema(
-                &self.project,
-                &self.data_source_id,
-                &self.table_id,
-                &table_schema,
-            )
-            .await?;
+        if !truncate {
+            // When doing incremental updates to a table's rows, the schema may become too wide.
+            // For example, if a column has only integers, it's an integer column. If a new row has a string in that column, the column becomes a string column.
+            // However, if that row is later updated to have an integer, the column should become an integer column again, but we cannot know that without looking at all the rows.
+            // This is why we invalidate the schema when doing incremental updates, and next time the schema is requested, it will be recomputed from all the rows.
+            store
+                .invalidate_table_schema(&self.project, &self.data_source_id, &self.table_id)
```

Comment:
Should we both mark the schema and update it?

Reply:
You are right. Updated (force-pushed)